core/cont/src/THashList.cxx

```diff
@@ -331,10 +331,22 @@ void THashList::RecursiveRemove(TObject *obj)
    }
 
    // Scan again the list and invoke RecursiveRemove for all objects
-   TIter next(this);
-   TObject *object;
-   while ((object = next())) {
-      if (object->TestBit(kNotDeleted)) object->RecursiveRemove(obj);
+   // We need to make sure to go through all the node even those
+   // marked as empty by another thread (Eventhough we hold the
+   // read lock if one of the call to RecursiveRemove request
+   // the write lock then the read lock will be suspended and
+   // another thread can modify the list; thanks to the shared_pointer
+   // forward-and-backward links, our view of the list is still intact
+   // but might contains node will nullptr payload)
```

Comment:
> Eventhough we hold the read lock if one of the call to RecursiveRemove request the write lock then the read lock will be suspended and another thread can modify the list

What? We cannot suspend the read lock. The writer *has* to wait for all readers to be done before writing anything.

Reply:
This is one of the case:

thread1: Take ReadLock; call Stuff; inside Stuff: Try to take the WriteLock
thread2: Take ReadLock; Call Stuff; inside Stuff: Try to take the WriteLock

there we can either have a dead-lock or one of the two threads must yield their read-lock ...

Reply:
Yes, readers cannot block after acquiring the lock for any reason. This means that you cannot "call stuff" after acquiring the read lock.

Reply:
Yes.  Although adhering strictly to this would mean excessive lock churn, hence the comment/behavior noted above.